File: src/services/order-service.js
const { Order, StateHistory } = require('../models/order');
const Recipe = require('../models/recipe');
const { StateMachine, OrderStates } = require('../core/state-machine');
const FruitPickerClient = require('./fruit-picker-client');
const BakerClient = require('./baker-client');
const DeliveryClient = require('./delivery-client');
const logger = require('../utils/logger');
const { v4: uuidv4 } = require('uuid');

class OrderService {
  constructor() {
    this.stateMachine = new StateMachine();
    this.fruitPickerClient = new FruitPickerClient();
    this.bakerClient = new BakerClient();
    this.deliveryClient = new DeliveryClient();
  }

  async createOrder(orderData) {
    const validPieTypes = ['apple', 'cherry', 'pumpkin', 'pecan', 'blueberry'];
    
    if (!validPieTypes.includes(orderData.pieType)) {
      throw new Error('Invalid pie type');
    }

    const recipe = await Recipe.findById(orderData.pieType);
    if (!recipe) {
      throw new Error('Recipe not found');
    }

    const estimatedDelivery = new Date();
    estimatedDelivery.setHours(estimatedDelivery.getHours() + 2);

    const order = await Order.create({
      id: uuidv4(),
      pieType: orderData.pieType,
      customerName: orderData.customer.name,
      customerEmail: orderData.customer.email,
      customerPhone: orderData.customer.phone,
      deliveryStreet: orderData.deliveryAddress.street,
      deliveryCity: orderData.deliveryAddress.city,
      deliveryState: orderData.deliveryAddress.state,
      deliveryZip: orderData.deliveryAddress.zip,
      currentState: OrderStates.ORDERED,
      estimatedDelivery: estimatedDelivery
    });

    await StateHistory.create({
      orderId: order.id,
      fromState: null,
      toState: OrderStates.ORDERED,
      notes: 'Order created'
    });

    logger.info(`Order created: ${order.id}`);

    this.processOrder(order.id).catch(err => {
      logger.error(`Failed to process order ${order.id}: ${err.message}`);
    });

    return order;
  }

  async getOrder(orderId) {
    const order = await Order.findByPk(orderId, {
      include: [{
        model: StateHistory,
        as: 'history',
        order: [['timestamp', 'ASC']]
      }]
    });

    if (!order) {
      throw new Error('Order not found');
    }

    return order;
  }

  async listOrders() {
    const orders = await Order.findAll({
      order: [['createdAt', 'DESC']],
      limit: 100
    });
    return orders;
  }

  async processOrder(orderId) {
    let order = await Order.findByPk(orderId);
    
    if (order.currentState === OrderStates.ORDERED) {
      await this.transitionState(order, OrderStates.PICKING);
      order = await Order.findByPk(orderId);
      
      const recipe = await Recipe.findById(order.pieType);
      const appleIngredient = recipe.ingredients.find(i => i.item === 'apples' || i.item === order.pieType + 's');
      const quantity = appleIngredient ? appleIngredient.quantity : 6;
      
      try {
        const pickJob = await this.fruitPickerClient.pickFruit(order.pieType, quantity);
        order.pickerJobId = pickJob.jobId;
        await order.save();
        
        await this.sleep(5000);
        
        await this.transitionState(order, OrderStates.PREPPING);
        order = await Order.findByPk(orderId);
      } catch (error) {
        await this.transitionState(order, OrderStates.ERROR, error.message);
        return;
      }
    }

    if (order.currentState === OrderStates.PREPPING) {
      await this.sleep(3000);
      
      await this.transitionState(order, OrderStates.BAKING);
      order = await Order.findByPk(orderId);
      
      const recipe = await Recipe.findById(order.pieType);
      
      try {
        const bakeJob = await this.bakerClient.scheduleBaking(
          order.pieType,
          recipe.bakingTemp,
          recipe.bakingTime
        );
        order.bakerJobId = bakeJob.jobId;
        await order.save();
        
        await this.sleep(8000);
        
        await this.transitionState(order, OrderStates.DELIVERING);
        order = await Order.findByPk(orderId);
      } catch (error) {
        await this.transitionState(order, OrderStates.ERROR, error.message);
        return;
      }
    }

    if (order.currentState === OrderStates.DELIVERING) {
      try {
        const deliveryJob = await this.deliveryClient.scheduleDelivery(
          {
            street: order.deliveryStreet,
            city: order.deliveryCity,
            state: order.deliveryState,
            zip: order.deliveryZip
          },
          {
            type: 'pie',
            size: 'medium',
            temperature: 'warm'
          }
        );
        order.deliveryId = deliveryJob.deliveryId;
        await order.save();
        
        await this.sleep(10000);
        
        await this.transitionState(order, OrderStates.COMPLETED);
      } catch (error) {
        await this.transitionState(order, OrderStates.ERROR, error.message);
        return;
      }
    }
  }

  async transitionState(order, newState, errorMessage = null) {
    const oldState = order.currentState;
    
    if (!this.stateMachine.canTransition(oldState, newState)) {
      throw new Error(`Cannot transition from ${oldState} to ${newState}`);
    }

    order.currentState = newState;
    await order.save();

    await StateHistory.create({
      orderId: order.id,
      fromState: oldState,
      toState: newState,
      errorMessage: errorMessage
    });

    logger.info(`Order ${order.id} transitioned: ${oldState} -> ${newState}`);
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = OrderService;

File: src/services/order_service.py
from datetime import datetime, timedelta
from typing import Optional, List
import structlog
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, func

from src.models.order import Order, StateHistory
from src.models.recipe import PIE_CATALOG
from src.core.state_machine import OrderStateMachine, OrderState, InvalidTransitionError
from src.services.fruit_picker_client import FruitPickerClient
from src.services.baker_client import BakerClient
from src.services.delivery_client import DeliveryClient

logger = structlog.get_logger()

VALID_PIE_TYPES = list(PIE_CATALOG.keys())


class OrderNotFoundError(Exception):
    pass


class InvalidOrderError(Exception):
    pass


class OrderService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.fruit_picker = FruitPickerClient()
        self.baker = BakerClient()
        self.delivery = DeliveryClient()

    async def create_order(self, pie_type: str, customer: dict, delivery_address: dict) -> Order:
        if pie_type not in VALID_PIE_TYPES:
            raise InvalidOrderError(f"Pie type '{pie_type}' is not available")

        if not customer.get("name") or not customer.get("email"):
            raise InvalidOrderError("Customer name and email are required")

        estimated_delivery = datetime.utcnow() + timedelta(hours=3)

        order = Order(
            pie_type=pie_type,
            customer_name=customer["name"],
            customer_email=customer["email"],
            customer_phone=customer.get("phone"),
            delivery_street=delivery_address["street"],
            delivery_city=delivery_address["city"],
            delivery_state=delivery_address["state"],
            delivery_zip=delivery_address["zip"],
            current_state=OrderState.ORDERED.value,
            estimated_delivery=estimated_delivery,
        )

        self.db.add(order)
        await self.db.flush()

        history = StateHistory(
            order_id=order.id,
            from_state=None,
            to_state=OrderState.ORDERED.value,
            notes="Order created",
        )
        self.db.add(history)
        await self.db.commit()
        await self.db.refresh(order)

        logger.info("order_created", order_id=order.id, pie_type=pie_type)
        return order

    async def get_order(self, order_id: str) -> Order:
        result = await self.db.execute(select(Order).where(Order.id == order_id))
        order = result.scalar_one_or_none()
        if not order:
            raise OrderNotFoundError(f"Order '{order_id}' does not exist")
        return order

    async def list_orders(self, status: Optional[str] = None, pie_type: Optional[str] = None) -> List[Order]:
        query = select(Order)
        if status:
            query = query.where(Order.current_state == status)
        if pie_type:
            query = query.where(Order.pie_type == pie_type)
        query = query.order_by(Order.created_at.desc())
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def advance_order(self, order_id: str) -> Order:
        order = await self.get_order(order_id)
        recipe = PIE_CATALOG.get(order.pie_type, {})
        current_state = OrderState(order.current_state)
        sm = OrderStateMachine(current_state)

        if current_state == OrderState.ORDERED:
            fruit_type = order.pie_type
            quantity = len(recipe.get("ingredients", []))

            try:
                pick_response = self.fruit_picker.pick_fruit(fruit_type, quantity)
                order.picker_job_id = pick_response.get("jobId")
                sm.transition_to(OrderState.PICKING, order_id, "Started fruit picking")
                order.current_state = OrderState.PICKING.value
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.ORDERED.value,
                    to_state=OrderState.PICKING.value,
                    notes="Fruit picking started",
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order
            except Exception as e:
                logger.error("picking_failed", order_id=order_id, error=str(e))
                order.current_state = OrderState.ERROR.value
                order.error_message = str(e)
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.ORDERED.value,
                    to_state=OrderState.ERROR.value,
                    error_message=str(e),
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order

        elif current_state == OrderState.PICKING:
            try:
                sm.transition_to(OrderState.PREPPING, order_id, "Ingredients received, starting prep")
                order.current_state = OrderState.PREPPING.value
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.PICKING.value,
                    to_state=OrderState.PREPPING.value,
                    notes="Ingredients received",
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order
            except Exception as e:
                logger.error("prepping_transition_failed", order_id=order_id, error=str(e))
                raise

        elif current_state == OrderState.PREPPING:
            baking_temp = recipe.get("bakingTemp", 350)
            baking_time = recipe.get("bakingTime", 45)

            try:
                bake_response = self.baker.start_baking(order.pie_type, baking_temp, baking_time)
                order.baker_job_id = bake_response.get("jobId")
                sm.transition_to(OrderState.BAKING, order_id, "Baking started")
                order.current_state = OrderState.BAKING.value
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.PREPPING.value,
                    to_state=OrderState.BAKING.value,
                    notes=f"Baking at {baking_temp}F for {baking_time} minutes",
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order
            except Exception as e:
                logger.error("baking_failed", order_id=order_id, error=str(e))
                order.current_state = OrderState.ERROR.value
                order.error_message = str(e)
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.PREPPING.value,
                    to_state=OrderState.ERROR.value,
                    error_message=str(e),
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order

        elif current_state == OrderState.BAKING:
            window_start = datetime.utcnow()
            window_end = window_start + timedelta(hours=2)

            delivery_address = {
                "street": order.delivery_street,
                "city": order.delivery_city,
                "state": order.delivery_state,
                "zip": order.delivery_zip,
            }

            try:
                delivery_response = self.delivery.schedule_delivery(
                    order_id,
                    delivery_address,
                    window_start.isoformat() + "Z",
                    window_end.isoformat() + "Z",
                )
                order.delivery_id = delivery_response.get("deliveryId")
                order.estimated_delivery = datetime.fromisoformat(
                    delivery_response.get("eta", window_end.isoformat()).replace("Z", "")
                )
                sm.transition_to(OrderState.DELIVERING, order_id, "Pie picked up for delivery")
                order.current_state = OrderState.DELIVERING.value
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.BAKING.value,
                    to_state=OrderState.DELIVERING.value,
                    notes=f"Delivering via drone {delivery_response.get('droneId', 'unknown')}",
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order
            except Exception as e:
                logger.error("delivery_failed", order_id=order_id, error=str(e))
                order.current_state = OrderState.ERROR.value
                order.error_message = str(e)
                history = StateHistory(
                    order_id=order.id,
                    from_state=OrderState.BAKING.value,
                    to_state=OrderState.ERROR.value,
                    error_message=str(e),
                )
                self.db.add(history)
                await self.db.commit()
                await self.db.refresh(order)
                return order

        elif current_state == OrderState.DELIVERING:
            sm.transition_to(OrderState.COMPLETED, order_id, "Delivery confirmed")
            order.current_state = OrderState.COMPLETED.value
            history = StateHistory(
                order_id=order.id,
                from_state=OrderState.DELIVERING.value,
                to_state=OrderState.COMPLETED.value,
                notes="Order delivered successfully",
            )
            self.db.add(history)
            await self.db.commit()
            await self.db.refresh(order)
            return order

        else:
            raise InvalidOrderError(f"Order in terminal state: {current_state.value}")

    async def get_order_count_by_status(self) -> dict:
        result = await self.db.execute(
            select(Order.current_state, func.count(Order.id)).group_by(Order.current_state)
        )
        rows = result.all()
        return {row[0]: row[1] for row in rows}
